### PR TITLE
Add support for AOVs to Python AD integrators.

### PR DESCRIPTION
**Pull request by dvicini**
## Description
 PR to add AOV support to Python AD integrators. The motivation for this is to be able to output arbitrary auxiliary values from Python AD integrators. This can be useful to add additional supervision, e.g., to segmentation masks for inverse rendering.

## Checklist

- [x] My code follows the [style guidelines](https://mitsuba.readthedocs.io/en/latest/src/developer_guide.html#coding-style) of this project
- [x] My changes generate no new warnings
- [x] My code also compiles for `cuda_*` and `llvm_*` variants. If you can't test this, please leave below
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I cleaned the commit history and removed any "Merge" commits
- [x] I give permission that the Mitsuba 3 project may redistribute my contributions under the terms of its [license](https://github.com/mitsuba-renderer/mitsuba/blob/master/LICENSE)

### CODE REVIEW

**Comment by dvicini** [comment]
I think this is working. The failing tests are due to one of the tutorials needing to be updated. I will make a PR to the tutorials repo as well, but maybe let's first make sure that this PR is in good shape and the interface is the one we agree on.

File: src/python/python/ad/integrators/common.py [hunk 1]
```diff
@@ -721,45 +742,53 @@ def splatting_and_backward_gradient_image(value: mi.Spectrum,
                 with dr.suspend_grad(pos, ray, weight):
                     L = dr.full(mi.Spectrum, 1.0, dr.width(ray))
                     dr.enable_grad(L)
-
+                    aovs = []
+                    for _ in self.aov_names():
+                        aov = dr.full(mi.Float, 1.0, dr.width(ray))
```
Comment (Speierers) [inline src/python/python/ad/integrators/common.py:747]:
```suggestion
                        aov = dr.ones(mi.Float, dr.width(ray))
```

File: src/python/python/ad/integrators/common.py [hunk 2]
```diff
@@ -721,45 +742,53 @@ def splatting_and_backward_gradient_image(value: mi.Spectrum,
...
                     splatting_and_backward_gradient_image(
                         value=L * weight,
                         weight=1.0,
-                        alpha=1.0
+                        alpha=1.0,
+                        aovs=[aov * weight for aov in aovs]
                     )
 
                     δL = dr.grad(L)
+                    δaovs = [dr.grad(aov) for aov in aovs]
```
Comment (Speierers) [inline src/python/python/ad/integrators/common.py:758]:
```suggestion
                    δaovs = dr.grad(aovs)
```

File: src/render/python/integrator_v.cpp [hunk 3]
```diff
@@ -247,28 +247,28 @@ MI_VARIANT class PyADIntegrator : public CppADIntegrator<Float, Spectrum> {
...
+                "δaovs"_a=py::none(),
                 "state_in"_a=py::none(),
                 "active"_a=active
             );
             // Third output is the ADIntegrator's state, which we ignore here
-            auto [spec, mask, _] =
+            auto [spec, mask, aovs_, _] =
                 sample_override(**kwargs).template cast<PyReturn>();
-
+            std::copy(aovs_.begin(), aovs_.end(), aovs);
```
Comment (Speierers) [inline src/render/python/integrator_v.cpp:271]: Does this actually work? I suppose at this point `aovs_` is already converted to a `std::vector<Float>`?

In my case I was casting the variables explicitly:

```c++
    // Convert AOVs from python object to float array if needed
    if (py::isinstance<py::list>(outputs)) {
        auto list = outputs.template cast<py::list>();
        for (size_t i = 0; i < list.size(); ++i)
            aovs[i] = list[i].template cast<Float>();
    }
 ```

**Review by Speierers — COMMENTED** [review]
Those changes look good to me 🚀 

You already addressed what I mentioned in our discussion on Slack in `integrator_v.cpp` regarding the broken trampoline for Python AOV integrators.

File: src/render/python/integrator_v.cpp (hunk 3 above)
Comment (dvicini) [reply src/render/python/integrator_v.cpp:271]: Yes, I think so. I followed the already existing code from the PySamplingIntegrator. We are explicitly casting the return type to an std::vector. I think the code you wrote is marginally more efficient, as it does not require the temporary std::vector, but I would favor readability & simplicity here.

**Comment by dvicini** [comment]
Thanks for the feedback @Speierers ! I incorporated your suggestions

**Review by njroussel — APPROVED** [review]
Thanks for all the work!

It looked good, I pushed a new version with some minor changes. I'll take over the remaining changes if there any failing tests :+1:

**Comment by dvicini** [comment]
Thanks, appreciate it!

**Pull request merged**